qiskit/transpiler/passes/layout/sabre_pre_layout.py

```diff
@@ -0,0 +1,217 @@
+        """
+
+        self.target = target
+        self.coupling_map = coupling_map
+        self.max_distance = max_distance
+        self.error_rate = error_rate
+        self.max_trials_vf2 = max_trials_vf2
+        self.call_limit_vf2 = call_limit_vf2
+        self.improve_layout = improve_layout
+
+        if self.target is not None:
+            self.coupling_map = self.target.build_coupling_map()
+
+        super().__init__()
+
+    def run(self, dag):
+        """Run the SabrePreLayout pass on `dag`.
+
+        The discovered starting layout is written to the property set
+        value ``sabre_starting_layouts``.
+
+        Args:
+            dag (DAGCircuit): DAG to create starting layout for.
+        """
+
+        if self.coupling_map is None:
+            raise TranspilerError(
+                "SabrePreLayout requires either target or coupling_map to be provided."
+            )
+
+        starting_layout = None
+        cur_distance = 1
+        while cur_distance <= self.max_distance:
+            augmented_map, augmented_error_map = self._add_extra_edges(cur_distance)
+            pass_ = VF2Layout(
+                augmented_map,
+                seed=0,
+                max_trials=self.max_trials_vf2,
+                call_limit=self.call_limit_vf2,
+            )
+            pass_.property_set["vf2_avg_error_map"] = augmented_error_map
+            pass_.run(dag)
+
+            if "layout" in pass_.property_set:
+                starting_layout = pass_.property_set["layout"]
+                break
+
+            cur_distance += 1
+
+        if cur_distance > 1 and starting_layout is not None:
+            # optionally improve starting layout
+            if self.improve_layout:
+                starting_layout = self._minimize_extra_edges(dag, starting_layout)
+            # write discovered layout into the property set
+            if "sabre_starting_layouts" not in self.property_set:
+                self.property_set["sabre_starting_layouts"] = [starting_layout]
+            else:
+                self.property_set["sabre_starting_layouts"].append(starting_layout)
+
+    def _add_extra_edges(self, distance):
+        """Augments the coupling map with extra edges that connect nodes ``distance``
+        apart in the original graph. The extra edges are assigned errors allowing VF2
+        to prioritize real edges over extra edges.
+        """
+        nq = len(self.coupling_map.graph.node_indices())
+        augmented_coupling_map = CouplingMap()
+        augmented_coupling_map.graph = self.coupling_map.graph.copy()
+        augmented_error_map = ErrorMap(nq)
+
+        for (x, y) in itertools.combinations(self.coupling_map.graph.node_indices(), 2):
+            d = self.coupling_map.distance(x, y)
+            if 1 < d <= distance:
+                augmented_coupling_map.add_edge(x, y)
+                augmented_error_map.add_error((x, y), self.error_rate)
```

Comment:
I know we were talking offline about scaling the error rate based on the edge distance. Do you think doing something like:

```suggestion
                augmented_error_map.add_error((x, y), self.error_rate * distance)
```

Or were you thinking of something more complicated?

Reply:
I think something like this is a good idea, except that if `error_rate * distance` becomes larger than `1`, then the final fidelity will be `0`, regardless of the layout (because the scoring code multiplies all the fidelities). So either we can leave with that (in the worst case we will get a crappy starting layout that Sabre will eventually not pick), or to automatically decrease `error_rate` to something like `1 / (2 * distance)`, or to consider some `arctan`-like function `f(distance)` such that `f(1) = 0` (real edges have no error) and `f(infinity)=1` (100% error), and everything else would be in-between.

Reply:
I like the asymptotic error rate approach that seems like the right fit for this use case. Although, I'd probably just be lazy and do something like `min(0.9999, self.error_rate * distance)` to approximate it :)

Reply:
This actually works well in practice, though at the end I have gone with an also simple asymptotic approach which resembles score calculations in sabre ``error_rate(d) = 1 - ( (1 - self.error_rate)^d )`` (but interestingly  gives the same results as this "lazy" heuristic).